codegen/templates/http_client.tmpl

```diff
@@ -170,35 +177,39 @@ func (c *{{$clientName}}) {{title .Name}}(
 	switch res.StatusCode {
 		case {{.OKStatusCode.Code}}:
 			var responseBody {{unref .ResponseType}}
+			{{if isPointerType .ResponseType -}}
 			err = res.ReadAndUnmarshalBody(&responseBody)
+			{{else -}}
+			err = res.ReadAndUnmarshalNonPointerValue(&responseBody)
+			{{end -}}
 			if err != nil {
-				return nil, respHeaders, err
+				return defaultRes, respHeaders, err
 			}
 
 			{{- if .ResHeaderFields }}
 			// TODO(jakev): read response headers and put them in body
 			{{- end}}
 
-			return &responseBody, respHeaders, nil
+			return {{if isPointerType .ResponseType}}&{{end}}responseBody, respHeaders, nil
```

Comment:
> or should we just modify the response type with an `&` prefix?

I don't quite get the question, but I will try to clear the confusion: `.ResponseType` is a full type, meaning if it's a pointer type, it has the `*` prefix. `responseBody` has the type of `unref .ResponseType`, `unref` removes the `*` prefix if it is present, which is necessary so that we can pass `&responseBody` to the unmarshal function. And `{{if isPointerType .ResponseType}}&{{end}}` is the opposite of `unref`, we can make it a template function, but I found it not so necessary.